tests/nexus_workflow_test.go

```diff
@@ -732,6 +733,224 @@ func (s *NexusWorkflowTestSuite) TestNexusOperationAsyncCompletion() {
+	})
+	s.NoError(err)
+
+	run, err := s.SdkClient().ExecuteWorkflow(ctx, client.StartWorkflowOptions{
+		TaskQueue: taskQueue,
+	}, "workflow")
+	s.NoError(err)
+
+	completionWFType := "completion_wf"
+	completionWFID := testcore.RandomizeStr(s.T().Name())
+	completionWFTaskQueue := testcore.RandomizeStr(s.T().Name())
+	completionWFStartReq := &workflowservice.StartWorkflowExecutionRequest{
+		RequestId:          uuid.NewString(),
+		Namespace:          s.Namespace(),
+		WorkflowId:         completionWFID,
+		WorkflowType:       &commonpb.WorkflowType{Name: completionWFType},
+		TaskQueue:          &taskqueuepb.TaskQueue{Name: completionWFTaskQueue, Kind: enumspb.TASK_QUEUE_KIND_NORMAL},
+		Input:              nil,
+		WorkflowRunTimeout: durationpb.New(100 * time.Second),
+		Identity:           "test",
+	}
+
+	startLink := &commonpb.Link_WorkflowEvent{
+		Namespace:  s.GetNamespaceID(s.Namespace()),
+		WorkflowId: completionWFID,
+		Reference: &commonpb.Link_WorkflowEvent_EventRef{
+			EventRef: &commonpb.Link_WorkflowEvent_EventReference{
+				EventType: enumspb.EVENT_TYPE_WORKFLOW_EXECUTION_STARTED,
+			},
+		},
+	}
+
+	finishStartReqCh := make(chan struct{})
+	startFinishedCh := make(chan struct{})
+	// Special Nexus task poller to handle waiting until start request is finished.
+	go func() {
+		defer close(startFinishedCh)
+
+		res, err := s.GetTestCluster().FrontendClient().PollNexusTaskQueue(ctx, &workflowservice.PollNexusTaskQueueRequest{
+			Namespace: s.Namespace(),
+			Identity:  uuid.NewString(),
+			TaskQueue: &taskqueuepb.TaskQueue{
+				Name: taskQueue,
+				Kind: enumspb.TASK_QUEUE_KIND_NORMAL,
+			},
+		})
+		s.NoError(err)
+
+		start := res.Request.Variant.(*nexuspb.Request_StartOperation).StartOperation
+		s.Equal(op.Name(), start.Operation)
+
+		start.CallbackHeader[nexus.HeaderOperationID] = completionWFID
+		completionWFStartReq.CompletionCallbacks = []*commonpb.Callback{
+			{
+				Variant: &commonpb.Callback_Nexus_{
+					Nexus: &commonpb.Callback_Nexus{
+						Url:    start.Callback,
+						Header: start.CallbackHeader,
+					},
+				},
+			},
+		}
+
+		completionRun, err := s.FrontendClient().StartWorkflowExecution(ctx, completionWFStartReq)
+		s.NoError(err)
+		startLink.RunId = completionRun.RunId
+
+		// Delay finishing start request until after completion is received.
+		<-finishStartReqCh
+
+		_, err = s.GetTestCluster().FrontendClient().RespondNexusTaskCompleted(ctx, &workflowservice.RespondNexusTaskCompletedRequest{
+			Namespace: s.Namespace(),
+			Identity:  uuid.NewString(),
+			TaskToken: res.TaskToken,
+			Response: &nexuspb.Response{
+				Variant: &nexuspb.Response_StartOperation{
+					StartOperation: &nexuspb.StartOperationResponse{
+						Variant: &nexuspb.StartOperationResponse_AsyncSuccess{
+							AsyncSuccess: &nexuspb.StartOperationResponse_Async{
+								OperationId: completionWFID,
+							},
+						},
+					},
+				},
+			},
+		})
+		s.NoErrorf(err, "Duplicate start response should be ignored.")
```

Comment:
AFAIR using suite methods from background routines may panic. Don't remember the details though.